pkg/apiserver/rest/dualwriter_mode2.go

```
@@ -410,3 +416,200 @@ func enrichLegacyObject(originalObj, returnedObj runtime.Object) error {
+
+					log.Info("updating item on unified storage", "name", name)
+				} else {
+					log.Info("inserting item on unified storage", "name", name)
+				}
+
+				objInfo := rest.DefaultUpdatedObjectInfo(item.objLegacy, []rest.TransformFunc{}...)
+				res, _, err := storage.Update(ctx,
+					name,
+					objInfo,
+					func(ctx context.Context, obj runtime.Object) error { return nil },
+					func(ctx context.Context, obj, old runtime.Object) error { return nil },
+					true, // force creation
+					&metav1.UpdateOptions{},
+				)
+				if err != nil {
+					log.WithValues("object", res).Error(err, "could not update in storage")
+					syncErr++
+				} else {
+					syncSuccess++
+				}
+			}
+
+			// delete if object does not exists on legacy but exists on storage
+			if item.objLegacy == nil && item.objStorage != nil {
+				outOfSync++
+
+				ctx = request.WithRequestInfo(ctx, &request.RequestInfo{
+					APIGroup:  requestInfo.APIGroup,
+					Resource:  requestInfo.Resource,
+					Name:      name,
+					Namespace: requestInfo.Namespace,
+				})
+
+				log.Info("deleting item from unified storage", "name", name)
+
+				deletedS, _, err := storage.Delete(ctx, name, func(ctx context.Context, obj runtime.Object) error { return nil }, &metav1.DeleteOptions{})
+				if err != nil {
+					if !apierrors.IsNotFound(err) {
+						log.WithValues("objectList", deletedS).Error(err, "could not delete from storage")
+					}
+					syncErr++
+				} else {
+					syncSuccess++
+				}
+			}
+		}
+
+		everythingSynced = outOfSync == syncSuccess
+
+		metrics.recordDataSyncerOutcome(mode2Str, kind, everythingSynced)
+		metrics.recordDataSyncerDuration(err != nil, mode2Str, kind, startSync)
+
+		log.Info("finished syncing items", "items", len(itemsByName), "updated", syncSuccess, "failed", syncErr, "outcome", everythingSynced)
+	})
+
+	if err != nil {
+		log.Error(err, "Server lock for dualwriter mode 2 sync already exists")
+	}
+
+	return everythingSynced, nil
```

Is there a reason we log errors directly from this function and always return a nil error?

---

Oh, good catch! This is incorrect, we need to return an error if the Syncer fails to acquire a lock

---

Fixed